tools/test-proxy/Azure.Sdk.Tools.TestProxy/Store/GitStore.cs

```diff
@@ -1,37 +1,486 @@
 using System.IO;
-using Azure.Sdk.Tools.TestProxy.Common;
 using System.Net;
 using System;
+using System.Text.Json;
+using System.Threading.Tasks;
+using System.Diagnostics;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Linq;
+using Azure.Sdk.Tools.TestProxy.Common.Exceptions;
 
 namespace Azure.Sdk.Tools.TestProxy.Store
 {
+    public class DirectoryEvaluation
+    {
+        public bool IsRoot;
+        public bool IsGitRoot;
+        public bool AssetsJsonPresent;
+    }
+
+    /// <summary>
+    /// This class provides an abtraction for dealing with git assets that are stored in an external repository. An "assets.json" within a repo folder is used to inform targeting.
+    /// </summary>
     public class GitStore : IAssetsStore
     {
-        public void Push(string pathToAssetsJson, string contextPath) {
-            var config = ParseConfigurationFile(pathToAssetsJson);
+        private HttpClient httpClient = new HttpClient();
+        public GitProcessHandler GitHandler = new GitProcessHandler();
+        public string DefaultBranch = "main";
+        public string FileName = "assets.json";
+
+        public GitStore() { }
+
+        public GitStore(GitProcessHandler processHandler) {
+            GitHandler = processHandler;
+        }
+
+        #region push, reset, restore, and other asset repo implementations 
+        /// <summary>
+        /// Pushes a set of changed files to the assets repo. Honors configuration of assets.json passed into it.
+        /// </summary>
+        /// <param name="pathToAssetsJson"></param>
+        /// <returns></returns>
+        public async Task Push(string pathToAssetsJson) {
+            var config = await ParseConfigurationFile(pathToAssetsJson);
+            var pendingChanges = DetectPendingChanges(config);
+            var onLatestSHA = true;
+
+            if (pendingChanges.Length > 0)
+            {
+                if(!GitHandler.TryRun(config, $"rev-parse origin/{config.AssetsRepoBranch}", out CommandResult result))
+                {
+                    // if we have a nonzero exit code, only code 128 is acceptable.
+                    if (result.ExitCode != 128)
+                    {
+                        throw GenerateInvokeException(result);
+                    }
+                }
+                else
+                {
+                    var retrievedSHA = result.StdOut.Trim();
+                    if (retrievedSHA != config.SHA)
+                    {
+                        onLatestSHA = false;
+                    }
+                }
+
+                if (onLatestSHA)
```

Comment:
This reads much better now! You could further simplify it by putting the if/else inside a single try/catch unless you plan to handle the exceptions differently.

Reply:
They aren't the same pattern of commands, so I'm ok with how it is right now!